File: Metro_Card/src/token_builder.py
from .base import BaseToken
from typing import List


class MetroCard(BaseToken):

    def __init__(self, id: str, balance: int):
        super().__init__(id, balance)

    def deduct(self, amount: int) -> bool:
        return super().deduct(amount)

    def isEnoughBalance(self, required: int) -> int:
        return super().isEnoughBalance(required)

    def topUp(self, amount: int):
        return super().topUp(amount)


class TokenBuilder:

    def __init__(self):
        self._token_pool = dict()

    def getAllTokens(self) -> List:
        return [token for token in self._token_pool.values()]

    def addToken(self, token: BaseToken) -> bool:
        if token.getId() in self._token_pool:
            return False
        self._token_pool[token.getId()] = token
        return True

    def getToken(self, token_id: str) -> BaseToken:
        return self._token_pool.get(token_id)

File: Metro_Card_Node/src/token_builder.js
const {
    BaseToken
} = require("./base")


class MetroCard extends BaseToken {
    /**
     * Metrocard constructor.
     * @param {Number.int} id
     * @param {String} name
     * @returns
     */
    constructor(id, balance) {
        super(id, balance);
    }

    /**
     * Deduct method deducts money from the card.
     * @param {Number.int} amount 
     * @returns 
     */
    deduct(amount) {
        return super.deduct(amount)
    }

    isEnoughBalance(required) {
        return super.isEnoughBalance(required)
    }

    topUp(amount) {
        return super.topUp(amount)
    }

}


class TokenBuilder {
    #token_pool //

    constructor() {
        this.#token_pool = new Map()
    }

    getAllTokens() {
        let tokens = new Array.from(this.#token_poll.values())
        return tokens
    }

    addToken(token) {
        if (this.#token_pool.has(token))
            return
        
        let token = new MetroCard()
        this.#token_pool[token.getId()] = token
        return true
    }

    getToken(token_id) {
        return this.#token_pool.get(token_id)
    }
}

modules.exports = {
    TokenBuilder
}